onnxruntime/core/providers/cpu/ml/ml_common.h

```diff
@@ -234,125 +219,109 @@ static inline void multiclass_probability(int64_t classcount, const std::vector<
   }
 }
 
-static const float ml_sqrt2 = 1.41421356f;
-
-static inline float ComputeLogistic(float val) {
-  float v = 1 / (1 + std::exp(-std::abs(val)));
+// y = \frac{1}{1+e^{-x}} , x \in R
+template <typename T>
+inline T ComputeLogistic(T val) {
+  T v = 1 / (1 + std::exp(-std::abs(val)));
   return (val < 0) ? (1 - v) : v;
 }
 
-static inline float ComputeProbit(float val) {
-  return ml_sqrt2 * ErfInv(2 * val - 1);
+// It assumes val is in [0,1]
+template <typename T>
+inline T ComputeProbit(T val) {
+  return static_cast<T>(M_SQRT2) * ErfInv(2 * val - 1);
 }
 
 static inline float sigmoid_probability(float score, float proba, float probb) {
   float val = score * proba + probb;
-  return 1 - ComputeLogistic(val);  // ref: https://github.com/arnaudsj/libsvm/blob/eaaefac5ebd32d0e07902e1ae740e038eaaf0826/svm.cpp#L1818
+  //ref: https://github.com/arnaudsj/libsvm/blob/eaaefac5ebd32d0e07902e1ae740e038eaaf0826/svm.cpp#L1818
+  return 1 - ComputeLogistic(val);
 }
 
-static inline void ComputeSoftmax(std::vector<float>& values) {
-  std::vector<float> newscores;
-  // compute exp with negative number to be numerically stable
-  float v_max = -std::numeric_limits<float>::max();
-  for (float value : values) {
-    if (value > v_max)
-      v_max = value;
-  }
-  float this_sum = 0.f;
-  for (float value : values) {
-    float val2 = std::exp(value - v_max);
-    this_sum += val2;
-    newscores.push_back(val2);
-  }
-  for (int64_t k = 0; k < static_cast<int64_t>(values.size()); k++) {
-    values[k] = newscores[k] / this_sum;
-  }
+/**
+ * if scores.size() == 1 and post_transform != POST_EVAL_TRANSFORM::PROBIT and
+ * add_second_class is in [0, 3], output 2 value else output scores.size()
+ * values
+ */
+template <typename T>
+void write_scores(std::vector<T>& scores, POST_EVAL_TRANSFORM post_transform, int64_t write_index, Tensor* Z,
+                  int add_second_class) {
```

Comment:
Are there perf gains from the changes to this method as the new implementation is a lot harder to follow?